collector/lib/Logging.h

```diff
@@ -71,25 +77,79 @@ class LogMessage {
               << std::left << std::setw(LevelPaddingWidth) << GetLogLevelName(level_)
               << " " << std::put_time(nowTm, "%Y/%m/%d %H:%M:%S")
               << "] ";
+  }
 
-    if (throttled_) {
-      std::cerr << "[Throttled] ";
-    }
-
-    if (include_file_) {
+  void PrintFile() {
+    if (CheckLogLevel(LogLevel::DEBUG)) {
       const char* basename = strrchr(file_, '/');
-      if (!basename) {
+      if (basename == nullptr) {
         basename = file_;
       } else {
         ++basename;
       }
       std::cerr << "(" << basename << ":" << line_ << ") ";
     }
+  }
+
+ private:
+  const char* file_;
+  int line_;
+  LogLevel level_;
+};
+
+class LogHeader : public ILogHeader {
+ public:
+  LogHeader(const char* file, int line, LogLevel level)
+      : ILogHeader(file, line, level) {}
+
+  void PrintHeader() override {
+    PrintPrefix();
+    PrintFile();
+  }
+};
+
+class ThrottledLogHeader : public ILogHeader {
+ public:
+  ThrottledLogHeader(const char* file, int line, LogLevel level, std::chrono::duration<unsigned int> interval)
+      : ILogHeader(file, line, level), interval_(interval) {}
+
+  void PrintHeader() override {
+    PrintPrefix();
+
+    std::cerr << "[Throttled " << count_ << " messages] ";
+    count_ = 0;
+
+    PrintFile();
+  }
+
+  bool ShouldPrint() {
+    std::chrono::duration elapsed = std::chrono::steady_clock::now() - last_log_;
+    count_++;
```

Comment:
It might cause some surprise in the future (imagine someone else reshuffling this code), that `ShouldPrint` function actually increments the counter. Maybe it would be more obvious to introduce another method `Suppress`, that will increment the count? In that case the macro would look like:
```
if (log_level & cond & CLOG_VAR(stmt).ShouldPrint) {
    LogMessage;
} else {
    CLOG_VAR(stmt).Suppress();
}
```

Reply:
Yeah, I didn't really like this approach but I couldn't think of a better solution. I like your suggestion though, I'll add that.

Reply:
Ok, now I remember why this is a pain, the macro needs to end with `LogMessage(...)` so we can directly `<<` into it. I could do something similar to what you proposed by not using braces around the if/else statements, but that causes a bunch of warnings about a dangling else condition and it's overall ugly.

Best I can think of is renaming `ShouldPrint` to `Suppress`, turn the logic of the returned value around and use that, so at least it's a bit more obvious that `Suppress` is telling the macro to not output the log message and increment the count.

Reply:
Interesting. IIUC currently it does

```
if (log_level && cond && CLOG_VAR(stmt).ShouldPrint())
    LogMessage()
```

How does it return `LogMessage(...)` if the if condition is false?

Reply:
It doesn't, it's a macro, so `LogMessage()` and everything to the `;` is skipped when the condition is false.

![image](https://github.com/user-attachments/assets/48b534f4-2eea-417e-a880-2764ca9f4e1e)

Reply:
Wouldn't then reversing the if condition solve this conundrum? Like

```
if (!log_level ||!cond || CLOG_VAR(stmt).ShouldPrint()) {
    Suppress()
 } else
    LogMessage(...)
```

Reply:
Problem with that is that missing the log level or having the condition be false would also add to the count of message, I think we need something like:
```cpp
if (log_level && cond) 
  if (!CLOG_VAR(stmt).ShouldPrint()) {
    Suppress()
  } else
    LogMessage(...)
```
But then you have a dangling else and the logic becomes difficult to follow.

You also get this really nice warning a bunch of times when compiling:
```
In file included from /home/mmoltras/go/src/github.com/stackrox/collector/collector/lib/system-inspector/EventExtractor.h:10,
                 from /home/mmoltras/go/src/github.com/stackrox/collector/collector/lib/ContainerMetadata.cpp:5:
/home/mmoltras/go/src/github.com/stackrox/collector/collector/lib/system-inspector/EventExtractor.h: In member function ‘const int64_t* collector::system_inspector::EventExtractor::get_event_rawres(sinsp_evt*)’:
/home/mmoltras/go/src/github.com/stackrox/collector/collector/lib/Logging.h:224:6: warning: suggest explicit braces to avoid ambiguous ‘else’ [-Wdangling-else]
  224 |   if (collector::logging::CheckLogLevel(collector::logging::LogLevel::lvl) && (cond)) \
      |      ^
/home/mmoltras/go/src/github.com/stackrox/collector/collector/lib/Logging.h:230:39: note: in expansion of macro ‘CLOG_THROTTLED_IF’
  230 | #define CLOG_THROTTLED(lvl, interval) CLOG_THROTTLED_IF(true, lvl, interval)
      |                                       ^~~~~~~~~~~~~~~~~
```

Reply:
Well, ok, then just go with whatever is easier + loud commentaries around to explain what this function does :)